src/Microsoft.DotNet.GenAPI/Tests/CSharpBuilderTests.cs

```diff
@@ -0,0 +1,150 @@
+using System;
+using System.Collections.Generic;
+using Microsoft.CodeAnalysis;
+using Microsoft.CodeAnalysis.CSharp;
+using Microsoft.DotNet.GenAPI.Shared;
+using Moq;
+using Xunit;
+
+namespace Microsoft.DotNet.GenAPI.Tests;
+
+public class CSharpBuilderTests
+{
+    [Fact]
+    public void BuildSimpleNamespacesTest()
+    {
+        var syntaxTree = @"
+namespace A
+{
+namespace B {}
+
+namespace C.D {}
+}
+";
+
+        var assembly = CompilationHelper.GetAssemblyFromSyntax(syntaxTree, enableNullable: false);
+
+        var orderProvider = new Mock<IAssemblySymbolOrderProvider>(MockBehavior.Strict);
+        var assemblySymbolFilter = new Mock<IAssemblySymbolFilter>(MockBehavior.Strict);
+        var syntaxWriter = new Mock<ISyntaxWriter>(MockBehavior.Strict);
+
+        orderProvider.Setup(o => o.Order(It.IsAny<IEnumerable<INamespaceSymbol>>())).Returns(
+            (IEnumerable<INamespaceSymbol> v) => { return v; });
+        orderProvider.Setup(o => o.Order(It.IsAny<IEnumerable<INamedTypeSymbol>>())).Returns(
+            (IEnumerable<INamedTypeSymbol> v) => { return v; });
+
+        assemblySymbolFilter.Setup(o => o.Include(It.IsAny<INamespaceSymbol>())).Returns(true);
+
+        var block = new Mock<IDisposable>();
+
+        syntaxWriter.Setup(o => o.WriteNamespace(new string[] { /* global namespace */ })).Returns(block.Object);
+        syntaxWriter.Setup(o => o.WriteNamespace(new string[] { "A" })).Returns(block.Object);
+        syntaxWriter.Setup(o => o.WriteNamespace(new string[] { "A", "B" })).Returns(block.Object);
+        syntaxWriter.Setup(o => o.WriteNamespace(new string[] { "A", "C" })).Returns(block.Object);
+        syntaxWriter.Setup(o => o.WriteNamespace(new string[] { "A", "C", "D" })).Returns(block.Object);
+
+        var builder = new CSharpBuilder(orderProvider.Object,
+            assemblySymbolFilter.Object, syntaxWriter.Object);
+        builder.WriteAssembly(assembly);
+    }
+
+    [Fact]
+    public void BuildSimpleStructureTest()
+    {
+        var syntaxTree = @"
+namespace A
+{
+    public struct PublicStruct { }
+    struct InternalStruct { }
+    readonly struct ReadonlyStruct { }
+    public readonly struct PublicReadonlyStruct { }
+    record struct RecordStruct { }
+    readonly record struct ReadonlyRecordStruct { }
+    public ref struct PublicRefStruct { }
+    public readonly ref struct PublicReadonlyRefStruct { }
+}
+";
```

Comment:
In C# 11, you can use Raw String Literals:

```suggestion
       var syntaxTree = """
		    namespace A
		    {
		        public struct PublicStruct { }
		        struct InternalStruct { }
		        readonly struct ReadonlyStruct { }
		        public readonly struct PublicReadonlyStruct { }
		        record struct RecordStruct { }
		        readonly record struct ReadonlyRecordStruct { }
		        public ref struct PublicRefStruct { }
		        public readonly ref struct PublicReadonlyRefStruct { }
		    }
			""";
```